File: LeetCode/LinkedList/python/removeDuplicatesFromSortedListII.py
from typing import Dict, List, Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def deleteDuplicates(self, head: Optional[ListNode]) -> Optional[ListNode]:

        dummy = ListNode(0, head)
        cur = dummy

        while cur is not None:
            if cur.next is not None and cur.next.next is not None and cur.next.val == cur.next.next.val:
                temp = cur.next.next
                
                while temp is not None and temp.next is not None and temp.val == temp.next.val:
                    temp = temp.next
                
                cur.next = temp.next
            else:
                cur = cur.next

        return dummy.next

    @staticmethod
    def testSolution(input: Dict[str, List[int] | int]) -> None:
        l1 = Solution.createLinkedList(input["llVals"])
        expected = Solution.createLinkedList(input["expected"])

        print("Input: head vals: " + Solution.linkedListValueToString(l1))
        print("Expected: " + Solution.linkedListValueToString(expected))
        
        res = Solution().deleteDuplicates(l1)

        print("Result: " + Solution.linkedListValueToString(res))
        print("PASS" if Solution.validateResult(res, expected) else "FAIL")
        print("-" * 50)

    @staticmethod
    def createLinkedList(headVals: List[int]) -> Optional[ListNode]:
        arLen = len(headVals)

        if arLen < 1:
            return None

        lNodesList = []

        for i in range(arLen):
            lNodesList.append(ListNode(headVals[i]))

            if i > 0:
                lNodesList[i - 1].next = lNodesList[i]

            if i == arLen - 1:
                lNodesList[i].next = None
        
        return lNodesList[0]

    @staticmethod
    def linkedListValueToString(ll: Optional[ListNode]) -> str:
        if ll == None:
            return "[ ]"

        s = "[ "

        while ll is not None:
            tmp = ll.val
            s += str(tmp)
            s += ", "
            ll = ll.next

        if len(s) > 0:
            s = s[0 : len(s) - 2]

        s += " ]"

        return s

    @staticmethod
    def validateResult(res: Optional[ListNode], expected: Optional[ListNode]) -> bool:
        if res is None and expected is None:
            return True

        resList = []
        expectedList = []

        while res is not None:
            resList.append(res.val)
            res = res.next

        while expected is not None:
            expectedList.append(expected.val)
            expected = expected.next

        return resList == expectedList

if __name__ == "__main__":
    input = {"llVals": [1,2,3,3,4,4,5], "expected": [1,2,5]}
    Solution.testSolution(input)

    input = {"llVals": [1,1,1,2,3], "expected": [2,3]}
    Solution.testSolution(input)

File: LeetCode/LinkedList/java/removeDuplicatesFromSortedListII.java
package LeetCode.LinkedList.java;

import java.util.ArrayList;
import java.util.List;

class ListNode {
    int val;
    ListNode next;

    ListNode() {
    }

    ListNode(int val) {
        this.val = val;
    }

    ListNode(int val, ListNode next) {
        this.val = val;
        this.next = next;
    }
}

record removeDuplicatesFromSortedListIIRecord(int[] headVals, int[] expectedVals) {
}

class removeDuplicatesFromSortedListII {
    public ListNode deleteDuplicates(ListNode head) {
        ListNode dummy = new ListNode(0, head);
        ListNode cur = dummy;

        while (cur != null) {
            if (cur.next != null && cur.next.next != null && cur.next.val == cur.next.next.val) {
                // We found some duplicates, now we remove all the duplicates
                ListNode temp = cur.next.next;

                while (temp != null && temp.next != null && temp.val == temp.next.val) {
                    temp = temp.next;
                }

                cur.next = temp.next;
            } else {
                // If there's no duplicates, move cur pointer to next node
                cur = cur.next;
            }
        }

        return dummy.next;
    }

    public static void main(String[] args) {
        removeDuplicatesFromSortedListIIRecord input = new removeDuplicatesFromSortedListIIRecord(
                new int[] { 1, 2, 3, 3, 4, 4, 5 },
                new int[] { 1, 2, 5 });
        removeDuplicatesFromSortedListII.testSolution(input);

        input = new removeDuplicatesFromSortedListIIRecord(
                new int[] { 1, 1, 1, 2, 3 },
                new int[] { 2, 3 });
        removeDuplicatesFromSortedListII.testSolution(input);

    }

    private static void testSolution(removeDuplicatesFromSortedListIIRecord input) {
        ListNode l1 = createLinkedList(input.headVals());
        ListNode expected = createLinkedList(input.expectedVals());

        System.out.println("Input: head vals: " + linkedListValueToString(l1));
        System.out.println("Expected vals: " + linkedListValueToString(expected));

        ListNode res = new removeDuplicatesFromSortedListII().deleteDuplicates(l1);
        System.out.println("Result: " + linkedListValueToString(res));
        System.out.println(validateResults(res, expected) ? "PASS" : "FAIL");
        System.out.println("-".repeat(50));
    }

    private static boolean validateResults(ListNode res, ListNode expected) {
        if (res == null && expected == null) {
            return true;
        }

        List<Integer> res_list = new ArrayList<>();
        List<Integer> expected_list = new ArrayList<>();

        while (res != null) {
            res_list.add(res.val);
            res = res.next;
        }

        while (expected != null) {
            expected_list.add(expected.val);
            expected = expected.next;
        }

        return res_list.equals(expected_list);
    }

    private static String linkedListValueToString(ListNode ll) {
        if (ll == null) {
            return "[ ]";
        }

        StringBuilder sb = new StringBuilder();

        sb.append("[ ");

        while (ll != null) {
            sb.append(ll.val);
            sb.append(", ");

            ll = ll.next;
        }

        // Let's remove the last comma added
        if (sb.length() > 0) {
            sb.setLength(sb.length() - 2);
        }

        sb.append(" ]");

        return sb.toString();
    }

    private static ListNode createLinkedList(int[] headVals) {
        if (headVals.length < 1) {
            return null;
        }

        int arLen = headVals.length;

        ListNode[] lNodes1 = new ListNode[arLen];

        // Create the linked list and connect them
        for (int i = 0; i < arLen; i++) {
            lNodes1[i] = new ListNode(headVals[i]);

            // If this is not the first or last node, connect the node[i-1].next to node[i]
            if (i > 0) {
                lNodes1[i - 1].next = lNodes1[i];
            }

            // If this is the last node, connect to null
            if (i == arLen - 1) {
                lNodes1[i].next = null;
            }
        }

        // Return the head of the LL
        return lNodes1[0];
    }
}
